modules/pam_unix/passverify.c

```diff
@@ -76,9 +76,15 @@ PAMH_ARG_DECL(int verify_pwd_hash,
 
 	strip_hpux_aging(hash);
 	hash_len = strlen(hash);
+
+	/* The passed password is empty */
+	if (p && p[0] == '\0' && !nullok) {
+		retval = PAM_AUTH_ERR;
+	}
```

Comment:
Hmm, this assignment doesn't seem to have any effect as `retval` is getting assigned later in this function anyway.

Reply:
Great catch! Forgot to test the `nullok == 0` case and missed it. Updated it to reside in `if {} else {}` branches.

Tested the cases with:

```c
               // empty request and entry, correct password:
               assert(verify_pwd_hash("", "", 1) == PAM_SUCCESS);
               assert(verify_pwd_hash("", "", 0) != PAM_SUCCESS);
               assert(verify_pwd_hash("", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 1) == PAM_SUCCESS);
               // this case:
               assert(verify_pwd_hash("", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 0) != PAM_SUCCESS);

               // empty request and entry, incorrect password:
               assert(verify_pwd_hash("1", "", 1) != PAM_SUCCESS);
               assert(verify_pwd_hash("1", "", 0) != PAM_SUCCESS);
               assert(verify_pwd_hash("1", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 1) != PAM_SUCCESS);
               assert(verify_pwd_hash("1", "$6$Yy4ty2jJ$bsVQWo8qlXC6UHq1/qTC3UR60ZJKmKApJ3Wj7DreAy8FxlVKtlDnplFQ7jMLVlDqordE7e4t49GvTb.aI59TP0", 0) != PAM_SUCCESS);

               // non-empty, correct password:
               assert(verify_pwd_hash("test", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 1) == PAM_SUCCESS);
               assert(verify_pwd_hash("test", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 0) == PAM_SUCCESS);
               // non-empty, incorrect password:
               assert(verify_pwd_hash("test2", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 1) != PAM_SUCCESS);
               assert(verify_pwd_hash("test2", "$y$j9T$BCdF/dSFuBgDI2aBHUCE3/$DsEklvSUTOp/sG2vtt0PHD1tKzHDIYClPO53Exd1vu1", 0) != PAM_SUCCESS);
```

The case 4 was broken, works now.